### ISSUE TITLE
Questions for community on namespace usage

### ISSUE THREAD

**Issue by KathleenDollard** [dotnet/vblang#351]
Could you let me know your thoughts on these: 

- Do you specify namespaces in the body of your files very often? (instead of the default in project properties)
- If yes, do you align your namespaces with your folder structure? 

Thanks for the input!!!

**Comment by KathleenDollard** [comment]
Here is some context...

https://github.com/dotnet/roslyn/issues/14341

My feeling is that this is going to just mess up VB users, but I want other people's opinion.

**Comment by pricerc** [comment]
@KathleenDollard 

Short answers to your questions:
1) yes
2) no

Comments:

1) Of course the namespace in a VB project isn't a *default* namespace, it's a *root* namespace. A subtle but important distinction. Having a root instead of default is fine as long as it is optional, and we can leave it blank if we want.  Personally, I'd rather have a *default* that gets applied to all new code files. It would allow more flexibility and (gasp!) easier conversion between VB and that-other-language, but I digress.

2) I'm not a fan of namespaces being forced to follow folder structures. I know that's a thing that people do, but it's a style choice. If you're going to do that then it becomes coding by convention, and then why bother declaring the namespaces at all, just name the folders?

I have worked on projects that supporters of that proposal would be horrified by, where the namespaces and folder structures were used for different purposes and *enforcing* that kind of namespace renaming would completely mess up the structure of those projects. E.g. I'm about to start some maintenance work on C# project where the folder structure is split up with things like 'Forms', 'UserControls' and Schemas; but within those folders, namespaces are used for things like 'Sales' and 'WorkInProgress', so that the Forms for `namespace.Sales` are in the 'Forms\Sales' folder, but the business logic in 'BO\Sales'. This allows different teams (UI and business logic) to work in different *folders* but in the same namespace.

**Comment by pricerc** [comment]
Actually, I just had another read of that proposal, and it looks like an awful amount of work to deal with something that a) you would hope is an uncommon exercise in the life cycle of a project, b) is probably not that hard to do with a folder rename and/or regex search-and-replace, and c) has more than a fighting chance of getting it wrong.

**Comment by genlu** [comment]
> Personally, I'd rather have a default that gets applied to all new code files. It would allow more flexibility and (gasp!) easier conversion between VB and that-other-language

@pricerc It's interesting that you mentioned this. It has been brought up during our discussion on the sync namespace/folder refactoring as well. Our hope is eventually we can provide support for default namespace for VB and clean up the mess/confusion around root vs default namespace.

FYI @jasonmalinowski @davkean

**Comment by KathleenDollard** [comment]
@genlu I think most VB programmers think that VB got this right. If there is ever a true default namespace for VB, that shouldn't replace the root namespace - and I don't think the default namespace would be used very often in VB. 

IOW, I don't think VB programmers are confused or think this is a mess ;-)

**Comment by reduckted** [comment]
>  I don't think VB programmers are confused or think this is a mess

Agreed. And if you do want to define something outside of the root namespace, you can always start the namespace with `Global.`.

```vb
Namespace Global.Foo
    Public Class Bar
    End Class
End Namespace
```

The full name of `Bar` is `Foo.Bar` regardless of what the root namespace is set to.

**Comment by genlu** [comment]
>  I don't think VB programmers are confused or think this is a mess

That I agree. There's no default namespace in VB and no root namespace in C#, so definitely no confusing for a specific language. What I meant is the way those two terms are often used in some context as if they are interchangeable. For example, default namespace is the value define by "rootnamespace" msbuild property.

**Comment by genlu** [comment]
> And if you do want to define something outside of the root namespace

@reduckted Is this a common thing people would do in VB?

**Comment by pricerc** [comment]
@genlu not common, but it happens.

Years ago, when coding for the .NET Compact Framework, I had a library shared between desktop and mobile, and there were some functions that I was using on the desktop in the System namespace, that were missing in the Compact Framework. There were a few ways I could have resolved it, in the end, I chose to add the missing functions within my Windows Mobile library, but in the System namespace. When the Compact Framework was later expanded to include those functions, I just removed them from my library without needing to change any other code.

**Comment by reduckted** [comment]
> Is this a common thing people would do in VB?

I don't know about everyone else, but for me, I've never needed to do it.

**Comment by franzalex** [comment]
> 
> 
> > And if you do want to define something outside of the root namespace
> 
> @reduckted Is this a common thing people would do in VB?

Yes, VB developers it's a common thing for VB developers who have to work with namespaces.

**Comment by franzalex** [comment]
>     * Do you specify namespaces in the body of your files very often? (instead of the default in project properties)

**As required by the project**
Not all my projects need multiple namespaces but in the ones that do, I specify namespaces in the body of the file that needs a different namespace.
 
>     * If yes, do you align your namespaces with your folder structure?

Not always; almost never.

**Comment by Nukepayload2** [comment]
> Do you specify namespaces in the body of your files very often? (instead of the default in project properties)

No, not very often. I explicitly specify namespaces in:
1. Medium or large projects, which contains more than 40 code files or have more than 6 `Module`s.
2. Generated projects, such as Windows Template Studio projects.
3. Translated projects. Most of them were C# projects.
4. Platform dependent projects. To make sure that unsafe codes (PInvoke, Marshal, CLR hacks, pointer algorithms) are called through wrappers.

> If yes, do you align your namespaces with your folder structure?

Usually. Unless there's a folder which name means target platform, or the namespace is too long. For example, in folder `Models\Core`, I'll use `Namespace Models.Core`. But in folder "Models\Style\Format\Formatters\Number\Interop\Windows", I'll choose `Namespace Models.Style.Format.Formatters` . Because the path is so long, and there is a folder named "Windows".

**Comment by rskar-git** [comment]
@KathleenDollard To directly answer your question, I've not been involved in any VB project that makes much use of namespaces beyond what VB provides by default. I have seen a few VB projects that do the namespace/directory alignment thing, but it seems to be occassional, and mostly when the project is meant to be a sort of "standard" library within an organization.

What follows is me prattling on, mostly as a help (I hope) for others who may be wondering what all this is about.

Alignment of a namespace with directory structure seems to be an idiom that arose as a means to keep large projects managable. If, in your project MyProj, class MyClass has its *.cs (or *.vb) file tucked within a subfolder foo, the convention is to then make its namespace "MyProj.foo", and therefore the full name of the class is "MyProj.foo.MyClass". Put another subfolder bar inside of foo, and another class MyOtherClass inside of bar, then, by that same convention, namespace is now "MyProj.foo.bar", and full name is "MyProj.foo.bar.MyOtherClass".

In Java this sort of alignment is mandatory. For that reason, it makes good sense that this would be a natural arrangement for C# too. Actually, in C#, the IDE automatically seeds a new class file with a `namespace` built from the directory structure and the "Default namespace" setting, which is (without a doubt) a hugh convenience at first...until, one day, of course, somebody sees the need to refactor the names (and/or structure) of the directories.

In VB, the very idea of specifying a namespace became a thing with the transition to .NET. Hence the extensive use of namespaces, let alone namespace/directory alignment, was and probably still is a novel thing. Perhaps it would be VB'ers who were also C/C++'ers (or the occassional VB+Java'er?) who would naturally think of organizing a project this way.

In any case, @genlu is right to say that discourse is muddled by the conflation of "root" versus "default". In VB, "Root namespace" really is the default when `Namespace` isn't explicitly given; and when it is explicitly given (at the top level), it is automatically prefixed behind-the-scenes (unless overridden with `Global` at the start). In C#, "Default namespace" is more of a setting to the IDE, so that when it generates the `namespace` statement automatically for a new class it will prefix the namespace with that setting; otherwise, without any `namespace` statement, the default is "global" (i.e. no behind-the-scenes prefixed root).

To put it another way, in C# a `namespace foo.bar` at the top is foo.bar - there is no implied root. In VB, `Namespace foo.bar` is <"Root namespace" setting>.foo.bar. If one prefers to be more C#-like in VB, "Root namespace" can be blank; in which case `Namespace foo.bar` is foo.bar. If "Root namespace" isn't blank, then `Namespace Global.foo.bar` gets one to foo.bar.

**Comment by jrmoreno1** [comment]
I seldom specify it, but when I do it, it is typically to match it to the folder structure.

But this brings to mind another issue -- why do namespaces and class names conflict?  It would seem that if I have a namespace "MyNameSpace.Space1.Space2" and a variable referencing a class named Space1 or Space2 it shouldn't be difficult to determine that Space1 is also a class as well as a namespace, and then use the class instead of giving the BC30182 error.
